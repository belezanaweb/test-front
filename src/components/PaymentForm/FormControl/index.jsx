import React from 'react'
import { useField } from '@unform/core'

import InputMask from './../FormInput'

import { Container, Label, ErrorMessage } from './styles'

function FormControl({ name, label, ...rest }) {
  const { error } = useField(name)

  return (
    <Container>
      {label && <Label>{label}</Label>}
      <InputMask name={name} {...rest} />
      {error && <ErrorMessage>{error}</ErrorMessage>}
    </Container>
  )
}

export default FormControl
