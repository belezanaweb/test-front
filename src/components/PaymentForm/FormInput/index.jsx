import React, { useRef, useEffect } from 'react'
import { useField } from '@unform/core'

import { StyledInput, StyledInputMask } from './styles'

function InputMask({ name, mask, ...rest }) {
  const inputRef = useRef(null)
  const { fieldName, registerField, defaultValue, error } = useField(name)

  useEffect(() => {
    registerField({
      name: fieldName,
      ref: inputRef.current,
      path: 'value',
      setValue(ref, value) {
        ref.setInputValue(value)
      },
      clearValue(ref) {
        ref.setInputValue('')
      }
    })
  }, [fieldName, registerField])

  if (mask) {
    return (
      <StyledInputMask
        ref={inputRef}
        defaultValue={defaultValue}
        error={error}
        alwaysShowMask={false}
        maskChar={null}
        mask={mask}
        {...rest}
      />
    )
  }

  return <StyledInput ref={inputRef} defaultValue={defaultValue} error={error} {...rest} />
}

export default InputMask
