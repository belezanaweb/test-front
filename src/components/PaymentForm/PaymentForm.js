import React, { forwardRef } from 'react'
import {
  PaymentFormSection,
  PaymentFormWrapper,
  PaymentFormWideInput,
  PaymentFormTitle
} from './PaymentForm.styles'
import { Input } from '../Input/Input'

export const PaymentForm = forwardRef(({ title = 'Cartão de Crédito', ...props }, ref) => (
  <PaymentFormSection>
    <PaymentFormTitle>{title}</PaymentFormTitle>
    <PaymentFormWrapper ref={ref} method="POST" id="_form" as="form" {...props}>
      <PaymentFormWideInput
        id="_input_credit_card"
        label="Número do cartão:"
        errorMessage="Número do cartão de crédito inválido"
        required
        name="cardnumber"
        autoComplete="cc-number"
        placeholder="____.____.____.____"
        pattern="\d{4}\.\d{4}\.\d{4}\.\d{4}"
        mask={[
          /\d/,
          /\d/,
          /\d/,
          /\d/,
          '.',
          /\d/,
          /\d/,
          /\d/,
          /\d/,
          '.',
          /\d/,
          /\d/,
          /\d/,
          /\d/,
          '.',
          /\d/,
          /\d/,
          /\d/,
          /\d/
        ]}
        guide
      />
      <PaymentFormWideInput
        id="_input_name"
        label="Nome do titular:"
        required
        name="ccname"
        autoComplete="cc-name"
        placeholder="Como no cartão"
      />
      <Input
        id="_input_expiry_date"
        label="Validade (mês/ano):"
        errorMessage="Validade do cartão de crédito inválida"
        required
        name="cc-exp"
        autoComplete="cc-exp"
        placeholder="__/____"
        pattern="\d{2}/\d{4}"
        mask={[/\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/]}
        guide
      />
      <Input
        id="_input_cvc"
        label="CVV:"
        errorMessage="CVV do cartão de crédito inválido"
        required
        name="cvc"
        autoComplete="cc-csc"
        placeholder="___"
        pattern="\d{3}"
        mask={[/\d/, /\d/, /\d/]}
        guide
      />
    </PaymentFormWrapper>
  </PaymentFormSection>
))
