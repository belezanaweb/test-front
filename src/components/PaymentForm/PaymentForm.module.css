.Form {
  display: flex;
  flex-direction: column;
}

.ContentTitle {
  margin-bottom: 16px;
}

.ShortCardInformations {
  display: flex;
  gap: 14px;
}