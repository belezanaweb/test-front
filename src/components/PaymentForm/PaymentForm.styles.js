import styled from 'styled-components'
import {
  GenericWrapperSection,
  GenericWrapperTitle,
  GenericWrapperWrapper
} from '../GenericWrapper/GenericWrapper'
import { Input } from '../Input/Input'

export const PaymentFormSection = styled(GenericWrapperSection)``

export const PaymentFormWrapper = styled(GenericWrapperWrapper)`
  display: grid;
  grid-template-columns: 52.45901% 1fr;
  grid-gap: 25px 20px;
`

export const PaymentFormWideInput = styled(Input)`
  grid-column-end: span 2;
`

export const PaymentFormTitle = styled(GenericWrapperTitle)``
