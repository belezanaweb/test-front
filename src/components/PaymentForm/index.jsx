import React from 'react'
import { SectionHeader } from '../SectionHeader'
import { SectionContainer, InnerContainer, DateCVVContainer, Input } from './styles'
import { ActionButton } from '../ActionButton'
import { Summary } from '../Summary'
import { useForm } from 'react-hook-form'
import { yupResolver } from '@hookform/resolvers/yup'
import * as yup from 'yup'

const paymentValidation = yup.object().shape({
  cardNumber: yup
    .string()
    .required('Campo inválido. Ex: 1234.1234.1234.1234')
    .min(19, 'Campo inválido. Ex: 1234.1234.1234.1234')
    .test('test-number', 'Campo inválido. Ex: 1234.1234.1234.1234', (value) =>
      /(\d{4}\.){3}/g.test(value)
    ),
  name: yup
    .string()
    .required('Campo inválido')
    .test(
      'test-name',
      'Campo inválido. No mínimo um nome e um sobrenome.',
      (value) => value.split(' ').length >= 2
    ),
  expiryDate: yup
    .string()
    .required('Campo inválido. Ex: 07/2024')
    .min(7, 'Campo inválido. Ex: 07/2024')
    .test('test-date', 'Campo inválido. Cartão vencido.', (value) =>
      /(0?[1-9]|1[0-2])\/(20[2-9][2-9]$)/.test(value)
    ),
  cvv: yup.string().required('Campo inválido').min(3, 'Campo inválido. Ex: 667')
})

export const PaymentForm = ({ handlePaymentInformation }) => {
  const {
    register,
    handleSubmit,
    getValues,
    formState: { errors, isValid }
  } = useForm({
    mode: 'all',
    resolver: yupResolver(paymentValidation)
  })

  const handleFormData = (data) => {
    handlePaymentInformation(data)
  }

  return (
    <>
      <SectionContainer>
        <SectionHeader text={'Cartão de Crédito'} />
        <InnerContainer>
          <form id="payment-form" data-testid="payment-form">
            <label htmlFor="cardNumber">
              <span>Número do Cartão:</span>
              <Input
                id="cardNumber"
                data-testid="card-number-input"
                type="text"
                placeholder="____.____.____.____"
                name="cardNumber"
                {...register('cardNumber')}
                maxLength={19}
                error={!getValues('cardNumber') || errors.cardNumber}
              />
              <p>{errors.cardNumber?.message}</p>
            </label>
            <label htmlFor="name">
              <span>Nome do Titular</span>
              <Input
                id="name"
                data-testid="name-input"
                type="text"
                placeholder="Como no cartão"
                name="name"
                {...register('name')}
                maxLength={50}
                error={!getValues('name') || errors.name}
              />
              <p>{errors.name?.message}</p>
            </label>
            <DateCVVContainer>
              <label htmlFor="expiryDate">
                <span>Validade (mês/ano):</span>
                <Input
                  id="expiryDate"
                  data-testid="expiry-date-input"
                  type="text"
                  placeholder="__/____"
                  name="expiryDate"
                  {...register('expiryDate')}
                  maxLength={7}
                  error={!getValues('expiryDate') || errors.expiryDate}
                />
                <p>{errors.expiryDate?.message}</p>
              </label>
              <label htmlFor="cvv">
                <span>CVV:</span>
                <Input
                  id="cvv"
                  data-testid="cvv-input"
                  type="text"
                  placeholder="___"
                  name="cvv"
                  {...register('cvv')}
                  maxLength={3}
                  error={!getValues('cvv') || errors.cvv}
                />
                <p>{errors.cvv?.message}</p>
              </label>
            </DateCVVContainer>
          </form>
        </InnerContainer>
      </SectionContainer>
      <SectionContainer>
        <Summary />
        <ActionButton
          type="submit"
          handleSubmit={handleSubmit(handleFormData)}
          form="payment-form"
          pathToGo={'/confirmation'}
          actionText={'Finalizar o pedido'}
          disabled={!isValid}
        />
      </SectionContainer>
    </>
  )
}
