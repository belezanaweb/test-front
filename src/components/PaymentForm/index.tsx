import React, { useState } from 'react'
import { Payment } from '../../types/productTypes'
import {
  FieldErrors,
  FieldValidate,
  PaymentValidate
} from '../../util/validations'
import Button from '../Button'
import InfoWrapper from '../InfoWrapper'
import Subtotal, { SubtotalProps } from '../Subtotal'
import TextField from '../TextField'
import Cards from 'react-credit-cards'
import * as S from './styles'
import 'react-credit-cards/es/styles-compiled.css'
import { Redirect } from 'react-router-dom'
import MediaMatch from '../MediaMatch'
import Dropdown from '../Dropdown'

type PaymentFormProps = {
  onSubmit: (value: Payment) => void
} & SubtotalProps

type focusCreditCard = 'name' | 'number' | 'expiry' | 'cvc' | undefined

const PaymentForm = ({
  shippingTotal,
  subTotal,
  total,
  discount,
  onSubmit
}: PaymentFormProps) => {
  const [fieldError, setFieldError] = useState<FieldErrors>({})
  const [focused, setFocused] = useState<focusCreditCard>(undefined)
  const [success, setSuccess] = useState(false)
  const [values, setValues] = useState<Payment>({
    creditCard: '',
    nameInCard: '',
    expirationDate: '',
    cvv: ''
  })

  const handleInput = (field: string, value: string) => {
    setValues((s) => ({ ...s, [field]: value }))
  }

  const handleOnBlur = (value: string, field: keyof Payment) => {
    let objError = fieldError
    const errors = FieldValidate(value, field)
    if (Object.keys(errors).length) {
      setFieldError({ ...fieldError, ...errors })
    } else {
      delete objError[field]
      setFieldError(objError)
    }
  }

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault()

    const errors = PaymentValidate(values)

    if (Object.keys(errors).length) {
      setFieldError(errors)
      return
    }
    onSubmit(values)
    setSuccess(true)
    setFieldError({})
  }

  return (
    <S.CartWrapper onSubmit={handleSubmit}>
      <S.BoxPaymentWrapper>
        <InfoWrapper title="CARTÃO DE CRÉDITO">
          <S.FormWrapper>
            <S.CardWrapper>
              <Cards
                cvc={values.cvv}
                expiry={values.expirationDate}
                focused={focused}
                name={values.nameInCard}
                number={values.creditCard}
              />
            </S.CardWrapper>
            <S.FormDetailsWrapper>
              <S.InputGroup isSideToSide={false}>
                <TextField
                  label={'Número do cartão:'}
                  name={'creditCard'}
                  type={'creditCard'}
                  error={fieldError?.creditCard}
                  onInputChange={(v) =>
                    handleInput(
                      'creditCard',
                      v
                        .split('')
                        .filter((char) => /^[0-9]*$/.test(char))
                        .join('')
                    )
                  }
                  handleOnFocus={() => setFocused('number')}
                  handleOnBlur={() => {
                    setFocused(undefined)
                    handleOnBlur(values.creditCard, 'creditCard')
                  }}
                  placeholder={'____.____.____.____'}
                  data-testid="creditCard"
                  mask={'9999.9999.9999.9999'}
                />

                <TextField
                  label={'Nome do Titular:'}
                  name={'nameInCard'}
                  type={'nameInCard'}
                  error={fieldError?.nameInCard}
                  onInputChange={(v) => handleInput('nameInCard', v)}
                  handleOnFocus={() => setFocused('name')}
                  handleOnBlur={() => {
                    setFocused(undefined)
                    handleOnBlur(values.nameInCard, 'nameInCard')
                  }}
                  placeholder={'Como no cartão'}
                  data-testid="nameInCard"
                  noMask={true}
                />
              </S.InputGroup>
              <S.InputGroup isSideToSide={true}>
                <TextField
                  name={'expirationDate'}
                  type={'expirationDate'}
                  error={fieldError?.expirationDate}
                  onInputChange={(v) =>
                    handleInput(
                      'expirationDate',
                      v
                        .split('')
                        .filter((char) => /^[0-9]*$/.test(char))
                        .join('')
                    )
                  }
                  label={'Validade (mês/ano):'}
                  handleOnFocus={() => setFocused('expiry')}
                  handleOnBlur={() => {
                    setFocused(undefined)
                    handleOnBlur(values.expirationDate, 'expirationDate')
                  }}
                  placeholder={'__/__'}
                  data-testid="expirationDate"
                  mask={'99/99'}
                />

                <TextField
                  name={'cvv'}
                  type={'cvv'}
                  error={fieldError?.cvv}
                  onInputChange={(v) =>
                    handleInput(
                      'cvv',
                      v
                        .split('')
                        .filter((char) => /^[0-9]*$/.test(char))
                        .join('')
                    )
                  }
                  handleOnFocus={() => setFocused('cvc')}
                  handleOnBlur={() => {
                    setFocused(undefined)
                    handleOnBlur(values.cvv, 'cvv')
                  }}
                  label={'CVV:'}
                  placeholder={'___'}
                  data-testid="cvv"
                  mask={'999'}
                />
              </S.InputGroup>
            </S.FormDetailsWrapper>
          </S.FormWrapper>
        </InfoWrapper>
      </S.BoxPaymentWrapper>
      <S.PaymentInfo>
        <MediaMatch greaterThan="medium">
          <S.AcessibilityWrapper>
            <Dropdown />
          </S.AcessibilityWrapper>
        </MediaMatch>
        <Subtotal
          shippingTotal={shippingTotal}
          subTotal={subTotal}
          total={total}
          discount={discount}
        />
        <Button
          disabled={
            !values.creditCard ||
            !values.nameInCard ||
            !values.expirationDate ||
            !values.cvv ||
            Object.keys(fieldError).length !== 0
          }
          type={'submit'}
        >
          FINALIZAR O PEDIDO
        </Button>
      </S.PaymentInfo>
      {success && <Redirect to="/success" />}
    </S.CartWrapper>
  )
}

export default PaymentForm
