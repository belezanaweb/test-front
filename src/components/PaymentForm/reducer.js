import {
  formatCardNumber,
  formatHolderName,
  formatExpirationDate,
  formatCVVCode,
} from './utils';

export const initialState = {
  cardNumber: '',
  cardHolderName: '',
  expirationDate: '',
  cvv: '',
  errors: {},
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'cardNumber':
      return {
        ...state,
        [action.field]: formatCardNumber(action.value),
        errors: {
          ...state.errors,
          [action.field]: undefined,
        },
      };
    case 'cardHolderName':
      return {
        ...state,
        [action.field]: formatHolderName(action.value),
        errors: {
          ...state.errors,
          [action.field]: undefined,
        },
      };
    case 'expirationDate':
      return {
        ...state,
        [action.field]: formatExpirationDate(action.value),
        errors: {
          ...state.errors,
          [action.field]: undefined,
        },
      };
    case 'cvv':
      return {
        ...state,
        [action.field]: formatCVVCode(action.value),
        errors: {
          ...state.errors,
          [action.field]: undefined,
        },
      };
    case 'updateErrors':
      return {
        ...state,
        errors: action.errors,
      };
    default:
      return state;
  }
};
