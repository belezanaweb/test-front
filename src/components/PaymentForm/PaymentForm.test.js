import React from 'react'
import { render } from '@testing-library/react'
import { PaymentForm } from './PaymentForm'

test('renders PaymentForm', () => {
  const { container } = render(<PaymentForm />)

  expect(container).toMatchSnapshot()
})
