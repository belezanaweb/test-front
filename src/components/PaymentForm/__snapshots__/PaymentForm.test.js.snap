// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PaymentForm 1`] = `
.c1 {
  padding: 12px 12px 13px 13px;
  border-radius: 3px;
  background-color: #fff;
  box-shadow: 1px 1px 5px 0 rgba(0,0,29,0.22);
}

.c0 {
  display: block;
  margin: 0 0 5px 13px;
  color: #999;
  font-size: 14px;
  font-weight: 700;
  line-height: 1.21428571428571;
  text-transform: uppercase;
}

.c3 {
  display: grid;
  grid-gap: 5px;
}

.c8 {
  display: none;
  margin-top: 3px;
  color: #f30;
  font-size: 12px;
  line-height: 1.166666667;
}

.c6 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  height: 45px;
  padding: 13px;
  border: 1px solid #e7e7e7;
  border-radius: 3px;
  background-color: #fff;
  font-size: 16px;
  line-height: 1.1875;
  box-shadow: inset 0 1px 2px 0 rgba(0,0,0,0.2);
}

.c6::-webkit-input-placeholder {
  color: #e0e7ee;
}

.c6::-moz-placeholder {
  color: #e0e7ee;
}

.c6:-ms-input-placeholder {
  color: #e0e7ee;
}

.c6::placeholder {
  color: #e0e7ee;
}

.c6:focus {
  border-color: #a43287;
}

.c6:invalid:not(:placeholder-shown) {
  border-color: #f30;
}

.c6:invalid:not(:placeholder-shown) + .c7 {
  display: block;
}

.c5 {
  display: block;
  margin-left: 2px;
  color: #ccc;
  font-size: 12px;
  font-weight: 700;
  line-height: 1.16666666666667;
}

.c2 {
  display: grid;
  grid-template-columns: 52.45901% 1fr;
  grid-gap: 25px 20px;
}

.c4 {
  grid-column-end: span 2;
}

<div>
  <section
    class=""
  >
    <h1
      class="c0"
    >
      Cartão de Crédito
    </h1>
    <form
      class="c1 c2"
      id="_form"
      method="POST"
    >
      <div
        class="c3 c4"
      >
        <label
          class="c5"
          for="_input_credit_card"
        >
          Número do cartão:
        </label>
        <input
          autocomplete="cc-number"
          class="c6"
          id="_input_credit_card"
          name="cardnumber"
          pattern="\\\\d{4}\\\\.\\\\d{4}\\\\.\\\\d{4}\\\\.\\\\d{4}"
          placeholder="____.____.____.____"
          required=""
          value=""
        />
        <div
          class="c7 c8"
        >
          Número do cartão de crédito inválido
        </div>
      </div>
      <div
        class="c3 c4"
      >
        <label
          class="c5"
          for="_input_name"
        >
          Nome do titular:
        </label>
        <input
          autocomplete="cc-name"
          class="c6"
          id="_input_name"
          name="ccname"
          placeholder="Como no cartão"
          required=""
          value=""
        />
        <div
          class="c7 c8"
        >
          Invalid Field
        </div>
      </div>
      <div
        class="c3"
      >
        <label
          class="c5"
          for="_input_expiry_date"
        >
          Validade (mês/ano):
        </label>
        <input
          autocomplete="cc-exp"
          class="c6"
          id="_input_expiry_date"
          name="cc-exp"
          pattern="\\\\d{2}/\\\\d{4}"
          placeholder="__/____"
          required=""
          value=""
        />
        <div
          class="c7 c8"
        >
          Validade do cartão de crédito inválida
        </div>
      </div>
      <div
        class="c3"
      >
        <label
          class="c5"
          for="_input_cvc"
        >
          CVV:
        </label>
        <input
          autocomplete="cc-csc"
          class="c6"
          id="_input_cvc"
          name="cvc"
          pattern="\\\\d{3}"
          placeholder="___"
          required=""
          value=""
        />
        <div
          class="c7 c8"
        >
          CVV do cartão de crédito inválido
        </div>
      </div>
    </form>
  </section>
</div>
`;
